swap/src/cli/cancel_and_refund.rs

```diff
@@ -55,26 +60,35 @@ pub async fn cancel(
 
     tracing::info!(%swap_id, "Manually cancelling swap");
 
-    let (txid, subscription) = match state6.submit_tx_cancel(bitcoin_wallet.as_ref()).await {
-        Ok(txid) => txid,
+    match state6.submit_tx_cancel(bitcoin_wallet.as_ref()).await {
+        Ok((txid, _)) => {
+            let state = BobState::BtcPunished {
+                tx_lock_id: state6.tx_lock_id(),
+            };
+            db.insert_latest_state(swap_id, state.clone().into())
+                .await?;
+            Ok((txid, state))
+        }
         Err(err) => {
             if let Ok(error_code) = parse_rpc_error_code(&err) {
                 tracing::debug!(%error_code, "parse rpc error");
-                if error_code == i64::from(RpcErrorCode::RpcVerifyAlreadyInChain) {
-                    tracing::info!("Cancel transaction has already been confirmed on chain");
-                } else if error_code == i64::from(RpcErrorCode::RpcVerifyError) {
-                    tracing::info!("General error trying to submit cancel transaction");
+                if error_code == i64::from(RpcErrorCode::RpcVerifyAlreadyInChain)
+                    || error_code == i64::from(RpcErrorCode::RpcVerifyError)
```

Comment:
Does `RpcErrorCode::RpcVerifyError` really always mean that the transaction has already been confirmed? What if the cancel timelock hasn't expired yet and we try to publish the cancel transaction?

We cannot move into the `BtcCancelled` state if that is the case.

Reply:
Oh, i think you are right.

Reply:
I added a timelock check, if I understand this part of the code correctly it should work fine now.